Source/AwsCommonRuntimeKit/io/Stream.swift

```diff
@@ -4,40 +4,36 @@
 import AwsCIo
 import Foundation
 
-// swiftlint:disable trailing_whitespace
+public protocol AwsStream {
+    var isEndOfStream: Bool { get }
+    var length: UInt { get }
+
+    func seek(offset: Int64, streamSeekType: StreamSeekType) -> Bool
+    /// Data.count should not greater than length.
+    func read(length: Int) -> Data
+}
 
-private var vtable = aws_input_stream_vtable(seek: doSeek,
-        read: doRead,
-        get_status: doGetStatus,
-        get_length: doGetLength,
-        acquire: { _ = Unmanaged<AwsInputStream>.fromOpaque($0!.pointee.impl).retain() },
-        release: { Unmanaged<AwsInputStream>.fromOpaque($0!.pointee.impl).release() })
+/// Direction to seek the stream.
+public enum StreamSeekType: UInt32 {
+    /// Seek the stream starting from beginning
+    case begin = 0
+    /// Seek the stream from End.
+    case end = 2
+}
 
 public class AwsInputStream {
-    var rawValue: aws_input_stream
-    let awsStream: AwsStream
     public var length: Int64
+    let awsInputStreamCore: AwsInputStreamCore
     public init(_ impl: AwsStream, allocator: Allocator = defaultAllocator) {
         length = Int64(impl.length)
-        awsStream = impl
-        rawValue = aws_input_stream()
-        rawValue.vtable = UnsafePointer<aws_input_stream_vtable>(&vtable)
-        rawValue.impl = Unmanaged<AwsInputStream>.passUnretained(self).toOpaque()
+        awsInputStreamCore = AwsInputStreamCore(awsStream: impl, allocator: allocator)
     }
 }
 
-public protocol AwsStream {
-    var status: aws_stream_status { get }
-    var length: UInt { get }
-
-    func seek(offset: Int64, basis: aws_stream_seek_basis) -> Bool
-    func read(buffer: inout aws_byte_buf) -> Bool
-}
-
 extension FileHandle: AwsStream {
     @inlinable
-    public var status: aws_stream_status {
-        return aws_stream_status(is_end_of_stream: self.length == self.offsetInFile, is_valid: true)
+    public var isEndOfStream: Bool {
+        self.length == self.offsetInFile
```

Comment:
under the hood `.length` is seeking to the end of the stream, that's an expensive isEndOfStream check 

in our other language bindings, in the read() functions, when we hit the end of the file we set some member variable like `self.endOfFile = true` ... and then later in the aws_input_stream_get_status() functions we report that we've hit the end

I see how you can't do that, due to this being a protocol and you can't add a member variable

maybe we need to update the protocol
remove `isEndOfStream`
add some way for `read()` to report when it's hit the end of the file, and we'll store that fact in InputStreamCore so we can report it later when `aws_input_stream_get_status()` is called

Reply:
I am not sure if Python binding has a bug. We are doing 
```
if (bytes_read == 0) {
    impl->is_end_of_stream = true;
} else {
    dest->len += bytes_read;
}
```
Let say stream has a length of 10. In the first read, we read 10 bytes. IsEnd should be true but it will be false until we do another read. 
I am thinking out about replacing isEndOfStream with currentOffset. and we can do a simple currentOffset == (pre cached) length check. Also can you please let me know what makes seeking to the end of a file expensive? I think, it should be O(1).

Reply:
Thank you for the detailed explanation. I have already updated this after discussion in the office.